// funky-shadow/dist/index.mjs
import { useRef, useMemo, useEffect } from 'react';
import { jsxs, jsx } from 'react/jsx-runtime';

// src/FunkyShadow.tsx

// src/presets.ts
var PRESETS = [
  { id: "crt-night", label: "CRT Night", stops: [[0, 0, 170], [85, 0, 255], [170, 0, 255], [255, 85, 255], [255, 215, 255], [255, 255, 255]] },
  { id: "thermal", label: "Thermal", stops: [[0, 0, 0], [0, 0, 135], [135, 0, 135], [255, 0, 0], [255, 175, 0], [255, 255, 85], [255, 255, 255]] },
  { id: "plasma", label: "Plasma", stops: [[0, 0, 135], [135, 0, 175], [255, 0, 85], [255, 175, 0], [255, 255, 85], [85, 255, 85], [0, 175, 175]] },
  { id: "deep-scan", label: "Deep Scan", stops: [[0, 0, 0], [0, 0, 95], [0, 0, 175], [0, 95, 255], [85, 175, 255], [175, 215, 255], [235, 245, 255]] },
  { id: "sunset-strip", label: "Sunset Strip", stops: [[0, 0, 55], [55, 0, 95], [175, 0, 95], [255, 55, 55], [255, 135, 55], [255, 215, 55], [255, 255, 175]] },
  { id: "vaporwave", label: "Vaporwave", stops: [[15, 0, 40], [55, 0, 135], [175, 0, 255], [255, 85, 215], [85, 255, 255], [255, 255, 255]] },
  { id: "galaxy", label: "Galaxy", stops: [[0, 0, 0], [15, 0, 55], [55, 0, 175], [135, 55, 255], [255, 135, 215], [255, 215, 255], [255, 255, 255]] },
  { id: "magma", label: "Magma", stops: [[0, 0, 0], [35, 0, 55], [135, 0, 95], [215, 55, 0], [255, 175, 0], [255, 255, 175]] },
  { id: "ocean", label: "Ocean", stops: [[0, 0, 0], [0, 0, 55], [0, 55, 135], [0, 135, 175], [85, 215, 215], [175, 255, 255]] },
  { id: "synthwave", label: "Synthwave", stops: [[0, 0, 35], [35, 0, 95], [135, 0, 175], [255, 0, 135], [255, 95, 55], [255, 215, 0]] },
  { id: "forest", label: "Forest", stops: [[0, 0, 0], [0, 35, 15], [0, 95, 35], [35, 135, 55], [95, 195, 95], [175, 235, 135]] },
  { id: "noir", label: "Noir", stops: [[0, 0, 0], [15, 15, 25], [55, 55, 75], [135, 135, 155], [215, 215, 235], [255, 255, 255]] }
];
function getPreset(id) {
  return PRESETS.find((p) => p.id === id);
}
function getPresetByIndex(index) {
  return PRESETS[index % PRESETS.length];
}

// src/types.ts
var DEFAULT_CONFIG = {
  offsetX: -15,
  offsetY: 20,
  blur: 30,
  opacity: 0.7,
  preset: "galaxy",
  shape: "radial",
  angle: 45,
  spread: 120,
  pixelScale: 3,
  dither: "4x4",
  quantLevels: 4,
  oklab: true,
  contrast: 130,
  brightness: 0
};

// src/render.ts
var BAYER2 = [[0, 2], [3, 1]];
var BAYER4 = [
  [0, 8, 2, 10],
  [12, 4, 14, 6],
  [3, 11, 1, 9],
  [15, 7, 13, 5]
];
var BAYER8 = [
  [0, 32, 8, 40, 2, 34, 10, 42],
  [48, 16, 56, 24, 50, 18, 58, 26],
  [12, 44, 4, 36, 14, 46, 6, 38],
  [60, 28, 52, 20, 62, 30, 54, 22],
  [3, 35, 11, 43, 1, 33, 9, 41],
  [51, 19, 59, 27, 49, 17, 57, 25],
  [15, 47, 7, 39, 13, 45, 5, 37],
  [63, 31, 55, 23, 61, 29, 53, 21]
];
function bayerThreshold(x, y, mode) {
  if (mode === "off") return 0.5;
  if (mode === "2x2") return (BAYER2[y % 2][x % 2] + 0.5) / 4;
  if (mode === "4x4") return (BAYER4[y % 4][x % 4] + 0.5) / 16;
  return (BAYER8[y % 8][x % 8] + 0.5) / 64;
}
function buildLevels(count) {
  if (count <= 1) return [0, 255];
  return Array.from({ length: count }, (_, i) => Math.round(i / (count - 1) * 255));
}
function ditherChannel(value, threshold, levels) {
  if (value <= levels[0]) return levels[0];
  if (value >= levels[levels.length - 1]) return levels[levels.length - 1];
  for (let i = 0; i < levels.length - 1; i++) {
    if (value >= levels[i] && value <= levels[i + 1]) {
      const lo = levels[i];
      const hi = levels[i + 1];
      return (value - lo) / (hi - lo) > threshold ? hi : lo;
    }
  }
  return value;
}
function lerpStops(stops, t) {
  if (t <= 0) return stops[0];
  if (t >= 1) return stops[stops.length - 1];
  const seg = t * (stops.length - 1);
  const i = Math.floor(seg);
  const f = seg - i;
  const a = stops[Math.min(i, stops.length - 1)];
  const b = stops[Math.min(i + 1, stops.length - 1)];
  return [a[0] + (b[0] - a[0]) * f, a[1] + (b[1] - a[1]) * f, a[2] + (b[2] - a[2]) * f];
}
function srgbToLinear(c) {
  const v = c / 255;
  return v <= 0.04045 ? v / 12.92 : Math.pow((v + 0.055) / 1.055, 2.4);
}
function linearToSrgb(c) {
  const v = c <= 31308e-7 ? c * 12.92 : 1.055 * Math.pow(c, 1 / 2.4) - 0.055;
  return Math.max(0, Math.min(255, v * 255));
}
function rgbToOklab(r, g, b) {
  const lr = srgbToLinear(r), lg = srgbToLinear(g), lb = srgbToLinear(b);
  const l = Math.cbrt(0.4122214708 * lr + 0.5363325363 * lg + 0.0514459929 * lb);
  const m = Math.cbrt(0.2119034982 * lr + 0.6806995451 * lg + 0.1073969566 * lb);
  const s = Math.cbrt(0.0883024619 * lr + 0.2817188376 * lg + 0.6299787005 * lb);
  return [
    0.2104542553 * l + 0.793617785 * m - 0.0040720468 * s,
    1.9779984951 * l - 2.428592205 * m + 0.4505937099 * s,
    0.0259040371 * l + 0.7827717662 * m - 0.808675766 * s
  ];
}
function oklabToRgb(l, a, b) {
  const l1 = l + 0.3963377774 * a + 0.2158037573 * b;
  const m1 = l - 0.1055613458 * a - 0.0638541728 * b;
  const s1 = l - 0.0894841775 * a - 1.291485548 * b;
  return [
    linearToSrgb(l1 * l1 * l1 * 4.0767416621 + m1 * m1 * m1 * -3.3077115913 + s1 * s1 * s1 * 0.2309699292),
    linearToSrgb(l1 * l1 * l1 * -1.2684380046 + m1 * m1 * m1 * 2.6097574011 + s1 * s1 * s1 * -0.3413193965),
    linearToSrgb(l1 * l1 * l1 * -0.0041960863 + m1 * m1 * m1 * -0.7034186147 + s1 * s1 * s1 * 1.707614701)
  ];
}
function lerpStopsOklab(stops, t) {
  if (t <= 0) return oklabToRgb(stops[0][0], stops[0][1], stops[0][2]);
  if (t >= 1) {
    const last = stops[stops.length - 1];
    return oklabToRgb(last[0], last[1], last[2]);
  }
  const seg = t * (stops.length - 1);
  const i = Math.floor(seg);
  const f = seg - i;
  const a = stops[i], b = stops[i + 1];
  return oklabToRgb(a[0] + (b[0] - a[0]) * f, a[1] + (b[1] - a[1]) * f, a[2] + (b[2] - a[2]) * f);
}
function gradientValue(x, y, w, h, shape, angle, spread) {
  const cx = w / 2, cy = h / 2;
  let t = 0;
  if (shape === "line") {
    const rad = angle * Math.PI / 180;
    const proj = (x - cx) * Math.cos(rad) + (y - cy) * Math.sin(rad);
    const maxProj = Math.abs(w / 2 * Math.cos(rad)) + Math.abs(h / 2 * Math.sin(rad));
    t = maxProj > 0 ? (proj + maxProj) / (2 * maxProj) : 0.5;
  } else {
    t = Math.sqrt(((x - cx) / (w / 2)) ** 2 + ((y - cy) / (h / 2)) ** 2) / Math.SQRT2;
  }
  t = 0.5 + (t - 0.5) * (spread / 100);
  return Math.max(0, Math.min(1, t));
}
function renderGradient(ctx, stops, rw, rh, shape, angle, spread, dither, quantLevels, oklab, contrast, brightness) {
  const levels = buildLevels(quantLevels);
  const contrastF = contrast / 100;
  const brightnessF = brightness * 2.55;
  const oklabStops = oklab ? stops.map(([r, g, b]) => rgbToOklab(r, g, b)) : null;
  const imgData = ctx.createImageData(rw, rh);
  const buf = imgData.data;
  for (let py = 0; py < rh; py++) {
    for (let px = 0; px < rw; px++) {
      const t = gradientValue(px, py, rw, rh, shape, angle, spread);
      let [r, g, b] = oklabStops ? lerpStopsOklab(oklabStops, t) : lerpStops(stops, t);
      r = Math.max(0, Math.min(255, (r - 128) * contrastF + 128 + brightnessF));
      g = Math.max(0, Math.min(255, (g - 128) * contrastF + 128 + brightnessF));
      b = Math.max(0, Math.min(255, (b - 128) * contrastF + 128 + brightnessF));
      const th = bayerThreshold(px, py, dither);
      const off = (py * rw + px) * 4;
      buf[off] = ditherChannel(r, th, levels);
      buf[off + 1] = ditherChannel(g, th, levels);
      buf[off + 2] = ditherChannel(b, th, levels);
      buf[off + 3] = 255;
    }
  }
  ctx.putImageData(imgData, 0, 0);
}
var BAYER8_MASK = [
  [0, 32, 8, 40, 2, 34, 10, 42],
  [48, 16, 56, 24, 50, 18, 58, 26],
  [12, 44, 4, 36, 14, 46, 6, 38],
  [60, 28, 52, 20, 62, 30, 54, 22],
  [3, 35, 11, 43, 1, 33, 9, 41],
  [51, 19, 59, 27, 49, 17, 57, 25],
  [15, 47, 7, 39, 13, 45, 5, 37],
  [63, 31, 55, 23, 61, 29, 53, 21]
];
function resolveStops(config) {
  if (config.colors) return config.colors;
  if (typeof config.preset === "string") {
    const found = getPreset(config.preset);
    return found ? found.stops : PRESETS[0].stops;
  }
  return getPresetByIndex(config.preset).stops;
}
function renderShadow(canvas, elementWidth, elementHeight, elementRadius, config = {}) {
  const cfg = { ...DEFAULT_CONFIG, ...config };
  const stops = resolveStops(cfg);
  const ctx = canvas.getContext("2d");
  if (!ctx) return;
  const ps = cfg.pixelScale;
  const blur = cfg.blur;
  const offX = cfg.offsetX;
  const offY = cfg.offsetY;
  const absOffX = Math.abs(offX);
  const absOffY = Math.abs(offY);
  const pad = blur * 2 + Math.max(absOffX, absOffY) + 20;
  const canvasW = elementWidth + pad * 2;
  const canvasH = elementHeight + pad * 2;
  const rw = Math.ceil(canvasW / ps);
  const rh = Math.ceil(canvasH / ps);
  canvas.width = rw;
  canvas.height = rh;
  const elW = elementWidth / ps;
  const elH = elementHeight / ps;
  const elR = elementRadius / ps;
  const elX = (rw - elW) / 2;
  const elY = (rh - elH) / 2;
  const sBlur = blur / ps;
  const sOffX = offX / ps;
  const sOffY = offY / ps;
  const mc = document.createElement("canvas");
  mc.width = rw;
  mc.height = rh;
  const mctx = mc.getContext("2d");
  const bigOff = rw + rh + sBlur * 4;
  mctx.shadowColor = "rgba(0,0,0,1)";
  mctx.shadowBlur = sBlur;
  mctx.shadowOffsetX = bigOff + sOffX;
  mctx.shadowOffsetY = sOffY;
  mctx.fillStyle = "rgba(0,0,0,1)";
  mctx.beginPath();
  mctx.roundRect(elX - bigOff, elY, elW, elH, elR);
  mctx.fill();
  const maskData = mctx.getImageData(0, 0, rw, rh).data;
  renderGradient(ctx, stops, rw, rh, cfg.shape, cfg.angle, cfg.spread, cfg.dither, cfg.quantLevels, cfg.oklab, cfg.contrast, cfg.brightness + Math.round((cfg.opacity * 100 - 50) * 0.5));
  const imageData = ctx.getImageData(0, 0, rw, rh);
  const data = imageData.data;
  for (let y = 0; y < rh; y++) {
    for (let x = 0; x < rw; x++) {
      const i = (y * rw + x) * 4;
      const ma = maskData[i + 3] / 255;
      if (ma === 0) {
        data[i + 3] = 0;
        continue;
      }
      const th = BAYER8_MASK[y % 8][x % 8] / 64;
      if (ma < th) {
        data[i] = 0;
        data[i + 1] = 0;
        data[i + 2] = 0;
        data[i + 3] = 0;
      } else {
        const l = (1 - ma) * 0.5;
        data[i] = Math.round(data[i] + (255 - data[i]) * l);
        data[i + 1] = Math.round(data[i + 1] + (255 - data[i + 1]) * l);
        data[i + 2] = Math.round(data[i + 2] + (255 - data[i + 2]) * l);
        data[i + 3] = 255;
      }
    }
  }
  ctx.putImageData(imageData, 0, 0);
}
function FunkyShadow({
  width,
  height,
  radius = 0,
  className,
  style,
  children,
  offsetX,
  offsetY,
  blur,
  opacity: opacityProp,
  preset,
  colors,
  shape,
  angle,
  spread,
  pixelScale,
  dither,
  quantLevels,
  oklab,
  contrast,
  brightness
}) {
  const canvasRef = useRef(null);
  const colorsKey = colors ? JSON.stringify(colors) : "";
  const config = useMemo(() => {
    const c = {};
    if (offsetX !== void 0) c.offsetX = offsetX;
    if (offsetY !== void 0) c.offsetY = offsetY;
    if (blur !== void 0) c.blur = blur;
    if (opacityProp !== void 0) c.opacity = opacityProp;
    if (preset !== void 0) c.preset = preset;
    if (colors !== void 0) c.colors = colors;
    if (shape !== void 0) c.shape = shape;
    if (angle !== void 0) c.angle = angle;
    if (spread !== void 0) c.spread = spread;
    if (pixelScale !== void 0) c.pixelScale = pixelScale;
    if (dither !== void 0) c.dither = dither;
    if (quantLevels !== void 0) c.quantLevels = quantLevels;
    if (oklab !== void 0) c.oklab = oklab;
    if (contrast !== void 0) c.contrast = contrast;
    if (brightness !== void 0) c.brightness = brightness;
    return c;
  }, [offsetX, offsetY, blur, opacityProp, preset, colorsKey, shape, angle, spread, pixelScale, dither, quantLevels, oklab, contrast, brightness]);
  useEffect(() => {
    const canvas = canvasRef.current;
    if (!canvas) return;
    renderShadow(canvas, width, height, radius, config);
  }, [width, height, radius, config]);
  const resolvedOpacity = opacityProp ?? DEFAULT_CONFIG.opacity;
  const cfg = { ...DEFAULT_CONFIG, ...config };
  const pad = cfg.blur * 2 + Math.max(Math.abs(cfg.offsetX), Math.abs(cfg.offsetY)) + 20;
  const canvasW = width + pad * 2;
  const canvasH = height + pad * 2;
  return /* @__PURE__ */ jsxs(
    "div",
    {
      className,
      style: {
        position: "relative",
        width,
        height,
        ...style
      },
      children: [
        /* @__PURE__ */ jsx(
          "canvas",
          {
            ref: canvasRef,
            style: {
              position: "absolute",
              width: canvasW,
              height: canvasH,
              left: `calc(50% - ${canvasW / 2}px)`,
              top: `calc(50% - ${canvasH / 2}px)`,
              pointerEvents: "none",
              imageRendering: "pixelated",
              opacity: resolvedOpacity
            }
          }
        ),
        /* @__PURE__ */ jsx("div", { style: { position: "relative", zIndex: 1, width: "100%", height: "100%" }, children })
      ]
    }
  );
}

export { DEFAULT_CONFIG, FunkyShadow, PRESETS, getPreset, getPresetByIndex, renderShadow };
